runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/metrics/MonitoringInfoMetricName.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.core.construction.metrics;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.base.Strings;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Objects;
+import org.apache.beam.sdk.metrics.MetricName;
+
+/**
+ * An implementation of {@code MetricKey} based on a MonitoringInfo's URN and label to represent the
+ * key instead of only a name+namespace. This is useful when defining system defined metrics with a
+ * specific urn via a {@code CounterContainer}.
+ */
+public class MonitoringInfoMetricName extends MetricName {
+
+  private String urn;
+  private String name;
+  private String namespace;
+  private HashMap<String, String> labels = new HashMap<String, String>();
+
+  public MonitoringInfoMetricName(String urn, HashMap<String, String> labels) {
+    checkArgument(!Strings.isNullOrEmpty(urn), "MonitoringInfoMetricName urn must be non-empty");
+    checkArgument(labels != null, "MonitoringInfoMetricName name must be non-null");
```

Comment:
Unclear comment.
Should state that labels parameter must not be null. If there are any required labels that should be present, it would be good to name those as well.